crypto/fipsmodule/cipher/e_aesccm.c

```diff
@@ -456,3 +494,239 @@ DEFINE_METHOD_FUNCTION(EVP_AEAD, EVP_aead_aes_128_ccm_matter) {
+    cipher_ctx->iv_set = 1;
+  }
+  return 1;
+}
+
+static int cipher_aes_ccm_cipher(EVP_CIPHER_CTX *ctx, uint8_t *out,
+                                 const uint8_t *in, size_t len) {
+  CIPHER_AES_CCM_CTX *cipher_ctx = CCM_CTX(ctx);
+  CCM128_CTX *ccm_ctx = CCM_INNER_CTX(cipher_ctx);
+  CCM128_STATE *ccm_state = CCM_INNER_STATE(cipher_ctx);
+
+  // Implicit EVP_*Final call. CCM does all the work in EVP_*Update
+  // n.b. |out| is non-NULL and |in| is NULL despite being a no-op.
+  if (in == NULL && out != NULL) {
+    return 0;
+  }
+
+  if (!cipher_ctx->iv_set || !cipher_ctx->key_set) {
+    return -1;
+  }
+
+  if (!out) {
+    if (!in) {
+      // If |out| and |in| are both NULL, |len| is the total length of the
+      // message which we need to include that in the 0th block of the CBC-MAC.
+      cipher_ctx->message_len = len;
+      cipher_ctx->len_set = 1;
+      return len;
+    } else {
+      // If only |out| is NULL then this is the AAD.
+      // The message length must be set apriori.
+      if (!cipher_ctx->len_set && len) {
+        return -1;
+      }
+      // We now have everything we need to initialize the CBC-MAC state
+      if (ccm128_init_state(ccm_ctx, ccm_state,
+                            &cipher_ctx->ks.ks, cipher_ctx->nonce,
+                            CCM_L_TO_NONCE_LEN(cipher_ctx->L), in, len,
+                            cipher_ctx->message_len)) {
+        cipher_ctx->ccm_set = 1;
+        return len;
+      } else {
+        return -1;
+      }
+    }
+  }
+
+  // The tag must be set before decrypting any data.
+  if (!EVP_CIPHER_CTX_encrypting(ctx) && !cipher_ctx->tag_set) {
+    return -1;
+  }
+  if (!cipher_ctx->len_set) {
+    return -1;
+  }
+  if (!cipher_ctx->ccm_set) {
+    // Initialize the ccm_state if this did not happen during the AAD update.
+    if (!ccm128_init_state(ccm_ctx, ccm_state, &cipher_ctx->ks.ks,
+                           cipher_ctx->nonce, CCM_L_TO_NONCE_LEN(cipher_ctx->L),
+                           NULL, 0, cipher_ctx->message_len)) {
+      return -1;
+    }
+    cipher_ctx->ccm_set = 1;
+  }
+
+  if (EVP_CIPHER_CTX_encrypting(ctx)) {
+    // Encryption path. Compute CBC-MAC on plaintext and then encrypt.
+    if (!ccm128_compute_mac(ccm_ctx, ccm_state, &cipher_ctx->ks.ks,
+                            cipher_ctx->tag, cipher_ctx->M, in, len)) {
+      return -1;
+    }
+    if (!ccm128_encrypt(ccm_ctx, ccm_state, &cipher_ctx->ks.ks, out, in, len)) {
+      return -1;
+    }
+    cipher_ctx->tag_set = 1;
+  } else {
+    // Decryption path. Compute the plaintext then compute its CBC-MAC.
+    // n.b. The method says encrypt, but it works both ways.
+    if (!ccm128_encrypt(ccm_ctx, ccm_state, &cipher_ctx->ks.ks, out, in, len)) {
+      return -1;
+    }
+    uint8_t computed_tag[EVP_AEAD_AES_CCM_MAX_TAG_LEN] = {0};
+    if (!ccm128_compute_mac(ccm_ctx, ccm_state, &cipher_ctx->ks.ks,
+                            computed_tag, cipher_ctx->M, out, len)) {
```

Comment:
should we `OPENSSL_cleanse(out, len)` before returning here as well?

Reply:
Yeah, that makes sense.